### PR TITLE
feat: rudderstackdataplane plugin implementation

### PR DESCRIPTION
**Pull request by SKannaniOS**
## Description
This PR introduces a dedicated plugin (`RudderStackDataPlanePlugin`)for managing message events and their operations.

## Implementation Details
This PR includes the following implementation
- `RudderStackDataPlanePlugin` - a plugin designed to handle messages through a proper flow.
- `AsyncChannel` - a  class utilizes `AsyncStream` to implement a subscription pattern, along with a dedicated serial queue.
- `MessageManager` - class performs the message operations like storage and upload to data plane

Note:
 - This PR provides a general Swift implementation related to the message handling, which will be adjusted in the future to meet the specific requirements for macOS, watchOS, and tvOS.
 
## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Code refactor/optimization

## Checklist

- [ ] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added the necessary documentation (if appropriate).
- [x] I have ensured that my code follows the project's code style.
- [x] I have checked for potential performance impacts and optimized if necessary.
- [x] I have checked the code for security issues.
- [ ] I have updated the changelog (if required).

### CODE REVIEW

File: Analytics/Analytics/Events/Message.swift [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+// MARK: - Message
+/**
+ This is base class for all events.
+ */
+public protocol Message: Codable {
+    var type: EventType { get set }
+    var messageId: String { get set }
+    var originalTimeStamp: String { get set }
+}
```
Comment (1abhishekpandey) [inline Analytics/Analytics/Events/Message.swift]: Let's consider adding `context` variable.

File: Analytics/Analytics/Plugins/RudderStackDataPlanePlugin.swift [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+//
+//  Created by Satheesh Kannan on 06/10/24.
+//
+
+import Foundation
+// MARK: - RudderStackDataPlanePlugin
+/**
+ This class serves as the main plugin responsible for initiating message operations.
+ */
+public final class RudderStackDataPlanePlugin: MessagePlugin {
```
Comment (1abhishekpandey) [inline Analytics/Analytics/Plugins/RudderStackDataPlanePlugin.swift]: Do we need this class to be `public`? Can we consider making it non-public?

File: Analytics/Analytics/Events/Message.swift [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+//  Created by Satheesh Kannan on 20/08/24.
+//
+
+import Foundation
+
+// MARK: - Message
+/**
+ This is base class for all events.
+ */
+public protocol Message: Codable {
```
Comment (1abhishekpandey) [inline Analytics/Analytics/Events/Message.swift]: Consider making it non-public.

File: Analytics/Analytics/Events/Message.swift [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    var type: EventType { get set }
+    var messageId: String { get set }
+    var originalTimeStamp: String { get set }
+}
+
+// MARK: - TrackEvent
+public struct TrackEvent: Message {
```
Comment (1abhishekpandey) [inline Analytics/Analytics/Events/Message.swift]: Consider making it non-public.

File: Analytics/Analytics/Events/Message.swift [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+        self.originalTimeStamp = .currentTimeStamp
+        
+        self.event = event
+        self.properties = properties
+        self.options = options
+    }
+}
+
+// MARK: - ScreenEvent
+public struct ScreenEvent: Message {
```
Comment (1abhishekpandey) [inline Analytics/Analytics/Events/Message.swift]: Consider making it non-public.

File: Analytics/Analytics/Events/Message.swift [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+        self.originalTimeStamp = .currentTimeStamp
+        
+        self.screenName = screenName
+        self.properties = properties
+        self.options = options
+    }
+}
+
+// MARK: - GroupEvent
+public struct GroupEvent: Message {
```
Comment (1abhishekpandey) [inline Analytics/Analytics/Events/Message.swift]: Consider making it non-public.

File: Analytics/Analytics/Events/Message.swift [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+        self.originalTimeStamp = .currentTimeStamp
+        
+        self.groupId = groupId
+        self.traits = traits
+        self.options = options
+    }
+}
+
+// MARK: - FlushEvent
+public struct FlushEvent: Message {
```
Comment (1abhishekpandey) [inline Analytics/Analytics/Events/Message.swift]: Consider making it non-public.

File: Analytics/Analytics/Plugins/Base/Plugin.swift [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+//  Analytics
+//
+//  Created by Satheesh Kannan on 20/08/24.
+//
+
+import Foundation
+
+// MARK: - PluginType
+@objc
+public enum PluginType: Int, CaseIterable {
```
Comment (1abhishekpandey) [inline Analytics/Analytics/Plugins/Base/Plugin.swift]: Consider making it non-public.

**Review by 1abhishekpandey — APPROVED** [review]

File: Analytics/Analytics/Events/Message.swift (hunk 1 above)
Comment (SKannaniOS) [reply Analytics/Analytics/Events/Message.swift]: I'll add it in the next PR.

**Review by 1abhishekpandey — APPROVED** [review]

**Review by ChryssaAliferi — APPROVED** [review]
Amazing! Let's go.

**Pull request merged**